### PR TITLE
Enable clangd's dot-to-arrow feature

### LINKED ISSUE #20815: Treat `.` as it was `->` for pointers and types with `operator->()` member functions
### Check for existing issues

- [X] Completed

### Describe the feature

Many IDEs and editors with C++ treat `.` as if it was `->`, depending on the type.

From the top of my head, there are 2 cases:
1. Variable of pointer type. Here `.` is a direct replacement for `->` because you cannot have `.` access on a variable of pointer type.
2. Variable of a type that implements `operator->()` member function. An example of this type would be `std::shared_ptr<T>` where `.` gives you access to the `std::shared_ptr` itself, but `->` gives you pointer to the underlying `T` type.

Activation of `->` entry should also replace `.` with `->`.

### Environment

Zed: v0.161.2 (Zed)
OS: Windows 10.0.19045
Memory: 31.9 GiB
Architecture: x86_64
GPU: NVIDIA GeForce GTX 1050 Ti || NVIDIA || 546.33

### If applicable, add mockups / screenshots to help present your vision of the feature

_No response_

### PR DESCRIPTION
**Pull request by feeiyu**
Closes #20815

dot2arrow1127

Release Notes:
- Enabled clangd's dot-to-arrow feature

### CODE REVIEW

**Review by SomeoneToIgnore — CHANGES_REQUESTED** [review]
1. Unfortunately, our CLA bot is broken, so I have to substitute for it for a while: 

> We require contributors to sign our Contributor License Agreement, and we don't have @feeiyu on file. You can sign our CLA at https://zed.dev/cla. Once you've signed, post a comment here that says '@cla-bot check'.

2. Seems that we do not need a custom request, type-less fiddling with `.` and a slight redesign of the approach to make it proper.

File: crates/lsp/src/lsp.rs [hunk 1]
```diff
@@ -319,6 +319,15 @@ impl lsp_types::notification::Notification for ServerStatus {
     const METHOD: &'static str = "experimental/serverStatus";
 }
 
+#[derive(Debug)]
+pub enum LspInitialize {}
```
[the remark this answers is not in the document]
Comment (feeiyu) [reply crates/lsp/src/lsp.rs:323]: yes, but it's not working for the old version of clangd if only fill the experimental field.  Should we consider compatibility?

[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply crates/lsp/src/lsp.rs:323]: https://github.com/zed-industries/zed/blob/2e21e7585a1aab6c34cc211463871b400c5541ab/crates/languages/src/c.rs#L42-L43

Zed always tries to install the latest stable clangd, so not sure it makes any sense.

File: crates/lsp/src/lsp.rs [hunk 2]
```diff
@@ -606,6 +615,7 @@ impl LanguageServer {
     pub fn initialize(
         mut self,
         options: Option<Value>,
+        callback: Option<Box<dyn FnOnce(&mut Value)>>,
```
[the remark this answers is not in the document]
Comment (feeiyu) [reply crates/lsp/src/lsp.rs:618]: The circular import happened after pass 'adapter: Option<&LspAdapter>' due to `LanguageServer` and `LspAdapter` in different crate `lsp` and `language`.
Maybe we can calculate the InitializeParams outside of initialize.

[the remark this answers is not in the document]
Comment (SomeoneToIgnore) [reply crates/lsp/src/lsp.rs:618]: Sure.
If nothing works, we can do the old way and pass two parameters, but let's avoid the closure that uses `Value`:
* we can either pass another `Option<Value>` for the experimental capabilities (not cool, but we can live with it and I may try and refactor it later)
* or follow your proposal indeed

**Review by SomeoneToIgnore — APPROVED** [review]
Looks much cleaner to me, thank you.

**Pull request merged**